### PR TITLE
Fix SDPA is_causal behavior to use top-left alignment

### PR DESCRIPTION
**Pull request by mangguo321**
### Details:
 - *Update CPU SDPA kernels to use top-left alignment for causal mask to align with SDPA specification*
 - *Keeps stateful KV-cache decode behavior correct by handling `fuse_concat && L0 > 0` separately*

### Tickets:
 - *CVS-189564*

### AI Assistance:
 - *AI assistance used: no / yes*
 - *If yes, summarize how AI was used and what human validation was performed (build/tests/manual checks).*

### CODE REVIEW

**Comment by maxnick** [comment]
@zhangYiIntel , could you please review?

File: src/plugins/intel_cpu/src/nodes/scaled_attn.cpp [hunk 1]
```diff
@@ -12,6 +12,7 @@
 #include <cstdint>
 #include <cstdlib>
 #include <cstring>
+#include <limits>
```
[the remark this answers is not in the document]
Comment (mangguo321) [reply src/plugins/intel_cpu/src/nodes/scaled_attn.cpp:15]: Done.

File: src/plugins/intel_cpu/src/nodes/scaled_attn.cpp [hunk 2]
```diff
@@ -1541,32 +1542,58 @@ struct ScaledDotProductAttention::AttentionExecutor : public ScaledDotProductAtt
...
-                        attn_mask =
-                            attn_mask.reshape({1, attn_mask.m_dims[0], attn_mask.m_dims[1], attn_mask.m_dims[2]});
+                // Speculative decode / chunked prefill with history (L1 > 1, L0 > 0):
+                // query row m is at position L0+m, attends to K[0..L0+m].
+                // Materialize the position-aware mask: bias[m,n] = 0 if n<=L0+m else -inf.
+                attn_buf.resize<T>({1, 1, L1, L0 + L1});
+                T* p = attn_buf.ptr<T>();
+                const T zero = static_cast<T>(0);
+                const T neg_inf = static_cast<T>(-std::numeric_limits<float>::infinity());
+                for (size_t m = 0; m < L1; m++) {
```
[the remark this answers is not in the document]
Comment (mangguo321) [reply src/plugins/intel_cpu/src/nodes/scaled_attn.cpp]: Done.

[the remark this answers is not in the document]
Comment (mangguo321) [reply src/plugins/intel_cpu/src/nodes/scaled_attn.cpp]: Done.

File: src/plugins/intel_cpu/src/nodes/scaled_attn.cpp [hunk 3]
```diff
@@ -1541,32 +1542,58 @@ struct ScaledDotProductAttention::AttentionExecutor : public ScaledDotProductAtt
...
-        } else {
-            if (is_causal) {
-                auto_causal = true;
+        } else if (is_causal && is_stateful_decode) {
+            if (L1 == 1) {
+                // Standard decode: the single new token is at absolute position L0, which is
+                // the last position in the cache — no future tokens exist. Full attention over
+                // all L0+1 keys is therefore correct and equal to the position-aware mask
+                // (every entry would be 0). Skip mask materialization entirely.
+                // Equivalent to PyTorch passing is_causal=False during decode.
```
[the remark this answers is not in the document]
Comment (mangguo321) [reply src/plugins/intel_cpu/src/nodes/scaled_attn.cpp:1557]: Done.

**Review by zhangYiIntel — APPROVED** [review]
LGTM, but I think we need to document the problem here.

## 1.  v13 SDPA defined in OV Spec uses top-left alignment when auto-casual is enabled
### Example: `kv_len = 8`, `q_len = 1`
```text
      K0 K1 K2 K3 K4 K5 K6 K7
Q0     ✓  ✗  ✗  ✗  ✗  ✗  ✗  ✗
```
Suitable for:
- Training
- Prefill
- Cases where `q_len == kv_len`
- NOT correct for decoding where q_len < kv_len

## 2. For decoding path, actually bottom-right alignment is correct. 
### Example: `kv_len = 8`, `q_len = 1`
```text
      K0 K1 K2 K3 K4 K5 K6 K7
Q0     ✓  ✓  ✓  ✓  ✓  ✓  ✓  ✓
```
@maxnick The solution in this PR is to use `bottom-right` for decoding when states are fused into SDPA while for none-stateful path uses top-left when `auto_casual` is enabled.

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/classes/concat_sdp.cpp [hunk 4]
```diff
@@ -198,7 +201,63 @@ void ConcatSDPTest::SetUp() {
...
+                                                 ref_beam_idx,
+                                                 ov::op::v0::Constant::create(ElementType::i32, {}, {0}));
+        auto ref_concatK = std::make_shared<ov::op::v0::Concat>(OutputVector{ref_gatherK, refParams[1]}, 2);
+        auto ref_concatV = std::make_shared<ov::op::v0::Concat>(OutputVector{ref_gatherV, refParams[2]}, 2);
+
+        auto ref_sdp = std::make_shared<ov::op::v13::ScaledDotProductAttention>(refParams[0],
+                                                                                ref_concatK,
+                                                                                ref_concatV,
+                                                                                ref_mask,
+                                                                                false);
```
Comment (zhangYiIntel) [inline src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/classes/concat_sdp.cpp:254]: why `causal` is not set to `true` ?

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/classes/concat_sdp.cpp [hunk 5]
```diff
@@ -198,7 +201,63 @@ void ConcatSDPTest::SetUp() {
     }
     SinkVector sinks{pastk_assign, pastv_assign};
     function = std::make_shared<ov::Model>(results, sinks, inputParams, "ConcatSDP");
-    functionRefs = function->clone();
+
+    if (m_isCausal) {
+        OPENVINO_ASSERT(!m_hasShapeOf, "Causal ConcatSDP reference does not support ShapeOf side outputs");
+        OPENVINO_ASSERT(m_headNumQ == m_headNumKV, "Causal ConcatSDP reference covers only non-GQA head layouts");
+        ov::ParameterVector refParams;
+        refParams.push_back(std::make_shared<ov::op::v0::Parameter>(inType, q_ps));
```
Comment (zhangYiIntel) [inline src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/classes/concat_sdp.cpp:214]: could we share the model construction code between `causal=True` & `causal=False`

**Review by zhangYiIntel — CHANGES_REQUESTED** [review]
Please fix the tests

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/classes/concat_sdp.cpp (hunk 5 above)
Comment (mangguo321) [reply src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/classes/concat_sdp.cpp:214]: For causal=False, the reference can reuse the original graph and then run ScaledDotProductAttentionDecomposition. For stateful-decode causal=true case, however, the reference needs an explicit attention_mask based on the accumulated history. Reusing the same path would make the reference rely on the implicit causal mask generated by ScaledDotProductAttentionDecomposition which is standard top-left form and does not capture the required history offset.

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/classes/concat_sdp.cpp (hunk 4 above)
Comment (mangguo321) [reply src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/classes/concat_sdp.cpp:254]: Casual is intentionally false because the reference encodes stateful-decode causality through the explicit attention_mask built in run_causal_reference(). The correct rule depends on past_len (n <= past_len + m), which SDPA's built-in is_causal does not represent here. Setting causal=true would apply an additional standard causal constraint and change the reference semantics.

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/classes/concat_sdp.cpp (hunk 5 above)
Comment (zhangYiIntel) [reply src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/classes/concat_sdp.cpp:214]: If this is the limitation in reference please add log for reminder. thanks!

**Review by zhangYiIntel — APPROVED** [review]

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/classes/concat_sdp.cpp (hunk 5 above)
Comment (mangguo321) [reply src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/classes/concat_sdp.cpp:214]: Done. Thanks!
